crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -68,6 +69,7 @@ pub enum NotifyStreamError {
 #[derive(Debug, Default, Clone)]
 struct WatchDescriptor {
     description: HashMap<PathBuf, HashMap<Option<String>, Vec<FileEvent>>>,
+    watch_discriptor_id: HashMap<c_int, String>,
```

Comment:
It would be best if only  one of these data structures existed, not both. The `description` data structure's goal is to describe a watched file or directory. 

It's composed of 
> {
>  "parent_path": {Option<"file_name">: [FileEvent]},
> } 

This was done to distinguish between watching a directory or watching a file, such that for watching a file it looks like:

> {
>  "parent_path": {**Some(**"file_name"**)**: [FileEvent]},
> } 

And watching a directory it looks like:

> {
> "parent_path": {**None**: [FileEvent]},
>} 


It was needed because the event id was not there. Now that we have this id, i think the data structure can be changed altogether and we can use the new one proposed and find a way of adding the vector (or set) of FileEvent(s).

I leave it up to you to decide the new data structure but here is an idea. 

```python 
{
    "c_int": ("full_path", [FileEvent_1, FileEvent_2]),
}
``` 
I think a data structure like this fully describes an inotify event and removes a lot of extra code lower down.

Reply:
I agree with @initard, one should have single structure that leverages the watch descriptor id that has been recently added by the `inotify` crate.

```
struct WatchDescriptor {
    description: HashMap<c_int, (PathBuf, Vec<FileEvent>)>,
}
```

Reply:
The above-mentioned structure does not fit here. So, using the below structure.

struct WatchDescriptor {
    description: HashMap<c_int, Vec<FileOrDir>>,
}
pub struct FileOrDir {
    pub dir_path: PathBuf,
    pub file_name: Option<String>,
    pub masks: Vec<FileEvent>,
}